### ISSUE TITLE
problems with special version-datas

### ISSUE THREAD

**Issue by afischer211** [ctron/package-drone#100]
I try to deploy an org.apache.servicemix-artifact (bundle-id: **org.apache.servicemix.bundles.saxon**) with the version: **9.6.0.5_1**
The artifact is uploaded successfully and shown in the channel-content, but can not be consumed by features or so. I also get a warning message in the console-log (no exception):
`
WARN  d.d.pm.p2.servlet.DownloadHandler [DownloadHandler.java:100] Artifact not found - name: org.apache.servicemix.bundles.saxon_9.6.0.5, version: 1, classifier: bundle
`
Is there a possible problem with **underscores** in the version-data?

**Comment by ctron** [comment]
I just checked and yes, there is a problem  ...

The OSGi specification allows the bundle symbolic name and the version to have the character "_". The P2 artifacts file defines a rule how to build an HTTP get url, which by default is `${repoUrl}/features/${id}_${version}.jar`

Now the `DownloadHandler` needs to find the appropriate OSGi bundle, and therefore splits the requested file name by the underscore. Assuming that the BSN may contain an underscore the version part if defined "non greedy".

So having a bundle named `a.b.c_d.e.f` with version `1.2.3.4_5_6` will result in the P2 file name `a.b.c_d.e.f_1.2.3.4_5_6.jar`. Since `d.e.f` of course could also be `1.2.3` there is now way to figure out which part belongs to the symbolic name and which to the version.

**Comment by ctron** [comment]
One possible solution (thus the separate comment) would be to change the URL pattern to something different. For example `${repoUrl}/plugins/${id}/${version}.jar`. (replacing the `_` with a `/`).

**Comment by ctron** [comment]
Btw .. congrats for :100:

 :+1:

**Comment by ctron** [comment]
Ok, the new URL template will be:

```xml
<rule filter="(&amp; (classifier=osgi.bundle))" output="${repoUrl}/plugins/${id}/${version}/${id}_${version}.jar"/>
<rule filter="(&amp; (classifier=binary))" output="${repoUrl}/binary/${id}/${version}/${id}_${version}"/>
<rule filter="(&amp; (classifier=org.eclipse.update.feature))" output="${repoUrl}/features/${id}/${version}/${id}_${version}.jar"/>
```

This allows for dedicated id, version and file names.

**Comment by afischer211** [comment]
Will you fix the problem in a version 0.9.2 or in 0.10.0?

**Comment by ctron** [comment]
Yes, this is the big question ... 

For making a 0.9.x release I would need to make a branch, since there are too many changes on the master now which will make up the 0.10.0 release. I am also still working on the LockManager issue, to create a better solution that what the small fix I made currently does. However I don't think I will get this done today.

So if you need this right away, I would make a branch and create a 0.9.2 release from that, cherry picking from the master. However this would steal a few hours from the other tasks.

**Comment by afischer211** [comment]
I can wait until tomorrow for an new version, but the problem with the version-datas is important for us. So if you need more time for the other topics, I would prefer a maintenance-release 0.9.2.

**Comment by ctron** [comment]
Ok ... so I will make a 0.9.2 version.

**Comment by ctron** [comment]
The signed 0.9.2 release file should be available in the Wuala folder.

**Comment by afischer211** [comment]
Ok, thanks for your work. I will download and test it soon.

**Comment by afischer211** [comment]
The problem is solved with the special release 0.9.2. Looking forward to the next releases... ;-)
